### PR TITLE
Feature extraction enabling

### PR DESCRIPTION
**Pull request by pi314ever**
# What does this PR do?

This PR enables BERT-based models for text feature extraction.
A change to `BertModel.forward` was made to avoid a NAN bug related to autocasting to BF16.

Models enabled:
- [Supabase/gte-small](https://huggingface.co/Supabase/gte-small)
- [thenlper/gte-small](https://huggingface.co/thenlper/gte-small)
- [thenlper/gte-base](https://huggingface.co/thenlper/gte-base)
- [thenlper/gte-large](https://huggingface.co/thenlper/gte-large)

## Before submitting
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Comment by pi314ever** [comment]
### Benchmark
Inputs:
```python
input_texts = [
    "what is the capital of China?",
    "how to implement quick sort in python?",
    "Beijing",
    "sorting algorithms"
]
```

Setup: 
- 500 sequential requests, timing only the `model(**batch_dict)` and the `average_pool` operations
- All times in milliseconds

Run Script: 
```sh
python run_example.py --model_name_or_path $MODEL_NAME --use_hpu_graphs --n_iterations 500
python run_example.py --model_name_or_path $MODEL_NAME --use_hpu_graphs --n_iterations 500 --bf16
```

#### `Supabase/gte-small`

| Performance | A100     | Gaudi2  |
| ---------------- | -------- | ------- |
| BF16             | 10.18650 | 0.68102 |
| FP32             | 15.84258 | 0.71221 |

#### `thenpler/gte-small`

| Performance | A100     | Gaudi2  |
| ---------------- | -------- | ------- |
| BF16             | 10.38441 | 0.67359 |
| FP32             | 12.66478 | 0.72647 |
#### `thenpler/gte-base`

| Performance | A100     | Gaudi2  |
| ----------- | -------- | ------- |
| BF16        | 10.2325  | 0.84994 |
| FP32        | 15.81192 | 0.88951 |

#### `thenpler/gte-large`

| Performance | A100     | Gaudi2  |
| ----------- | -------- | ------- |
| BF16        | 17.26550 | 1.88221 |
| FP32        | 24.39661 | 2.03020 |

**Comment by vidyasiv** [comment]
@pi314ever , thanks for your PR. I was able to verify and run the models.
Could you rebase your branch onto latest main and verify on SynapseAI v1.16.0? 
I see test failure on Gaudi2 host: `AssertionError: 1.1957454681396484 not less than 0.715071`

File: Makefile
Comment (vidyasiv) [inline Makefile:45]: My suggestion would be to add this under fast_tests but will defer to @regisss on this

File: examples/text-feature-extraction/run_example.py
Comment (vidyasiv) [inline examples/text-feature-extraction/run_example.py:1]: Suggestion: File name should be specific like ` run_feature_extraction.py` similar to how the other example scripts are named

File: examples/text-feature-extraction/run_example.py
Comment (vidyasiv) [inline examples/text-feature-extraction/run_example.py:124]:
```suggestion
    logger.info(f"Average time per iteration: {(end_time - start_time) * 1000 / args.n_iterations:.5f} ms")
```

**Comment by pi314ever** [comment]
@vidyasiv I am in the process of reproducing the performance issue on SynapseAI v1.16.0. I have addressed the other suggestions though.

File: examples/text-feature-extraction/README.md
Comment (vidyasiv) [inline examples/text-feature-extraction/README.md:18]: @pi314ever, another thing - we usually add the link for the original script that was adapted for Gaudi2 in the code file (run_feature_extraction) and the README. If you wrote the feature extraction sample entirely yourself, this can be ignored

File: tests/test_feature_extraction.py
Comment (vidyasiv) [inline tests/test_feature_extraction.py:29]: If this is the baseline for Gaudi2, I think there needs to be a check with environment variable GAUDI2_CI - see other tests for examples of this.

**Comment by pi314ever** [comment]
@vidyasiv I was not able to reproduce the performance issue on SynapseAI 1.16. Also, what is the process for obtaining the non-CI values?

**Comment by vidyasiv** [comment]
> @vidyasiv I was not able to reproduce the performance issue on SynapseAI 1.16. Also, what is the process for obtaining the non-CI values?

Thanks for checking. I ran again on a different host and it passes now.

File: tests/test_feature_extraction.py
Comment (vidyasiv) [reply tests/test_feature_extraction.py:29]: This is the last update needed for my approval then I will handover to @regisss for review.

Comment (pi314ever) [reply tests/test_feature_extraction.py:29]: @vidyasiv What is the standard process to obtain the latency measurements for non-GAUDI2_CI? I notice that some times are relaxed on GAUDI2_CI while others are tighter.

Comment (vidyasiv) [reply tests/test_feature_extraction.py:29]: When GAUDI2_CI is not set, the test is meant for Gaudi1. If you have access to Gaudi1 you can provide the numbers otherwise you can skip on your end. I am not aware of any standard for obtaining latency measurements otherwise. I assume they have to pass on the CI machines and the devs with access to push will update the values if needed.

**Review by libinta — COMMENTED** [review]
@pi314ever as you changed the bert model file, have you tested other bert ci test cases?

**Comment by pi314ever** [comment]
@libinta I have not yet. Where can I find the other BERT CI test cases? Is it in `tests/test_encoder_decoder.py`, or are there others?

**Comment by pi314ever** [comment]
@libinta There seems to be accuracy discrepancies for `Babelscape/mrebel-large` and `facebook/bart-large-cnn` that are present in the main branch as well though. I was able to verify that tests passed for all other tests in `make slow_tests_text_generation_example`.

**Review by vidyasiv — CHANGES_REQUESTED** [review]
Please update test file for GAUDI1/2 CI checks

**Comment by pi314ever** [comment]
@vidyasiv  I do not have access to a Gaudi1 machine. Is there anyone who can add the Gaudi1 CI to this PR?

**Comment by vidyasiv** [comment]
> @vidyasiv I do not have access to a Gaudi1 machine. Is there anyone who can add the Gaudi1 CI to this PR?

I sent you a patch, you can use that with `git am < gaudi1_tests_style.patch`

**Review by vidyasiv — APPROVED** [review]

**Review by regisss — APPROVED** [review]

**Pull request merged**
